persistence/sql/migrations/20241609000001000000_device_flow.cockroach.up.sql

```diff
@@ -0,0 +1,46 @@
+-- Migration generated by the command below; DO NOT EDIT.
+-- hydra:generate hydra migrate gen
+CREATE TABLE IF NOT EXISTS hydra_oauth2_device_auth_codes
+(
+    device_code_signature VARCHAR(255) NOT NULL,
+    user_code_signature   VARCHAR(255) NOT NULL,
+    request_id            VARCHAR(40)  NOT NULL,
+    requested_at          TIMESTAMP    NOT NULL DEFAULT NOW(),
+    client_id             VARCHAR(255) NOT NULL,
+    scope                 TEXT         NOT NULL,
+    granted_scope         TEXT         NOT NULL,
+    form_data             TEXT         NOT NULL,
+    session_data          TEXT         NOT NULL,
+    subject               VARCHAR(255) NOT NULL DEFAULT '',
+    device_code_active    BOOL         NOT NULL DEFAULT true,
+    user_code_state       SMALLINT     NOT NULL DEFAULT 0,
+    requested_audience    TEXT         NULL DEFAULT '',
+    granted_audience      TEXT         NULL DEFAULT '',
+    challenge_id          VARCHAR(40)  NULL,
+    expires_at            TIMESTAMP    NULL,
+    nid                   UUID         NULL,
+
+    FOREIGN KEY (client_id, nid) REFERENCES hydra_client(id, nid) ON DELETE CASCADE,
+    FOREIGN KEY (nid) REFERENCES networks(id) ON UPDATE RESTRICT ON DELETE CASCADE,
+    PRIMARY KEY (device_code_signature, nid)
+);
+
+CREATE INDEX hydra_oauth2_device_auth_codes_request_id_idx ON hydra_oauth2_device_auth_codes (request_id, nid);
+CREATE INDEX hydra_oauth2_device_auth_codes_client_id_idx ON hydra_oauth2_device_auth_codes (client_id, nid);
+CREATE INDEX hydra_oauth2_device_auth_codes_challenge_id_idx ON hydra_oauth2_device_auth_codes (challenge_id);
+CREATE UNIQUE INDEX hydra_oauth2_device_auth_codes_user_code_signature_idx ON hydra_oauth2_device_auth_codes (user_code_signature, nid);
```

Comment:
I only found queries for `device_code_signature` (primary key) and `request_id)` - do we need the challenge-id?

Also are we not missing a foreign key onto the oauth2 flow table (I'm not sure if we have that in the other tables)

Since you're referencing `nid` as a foreign key, it will also need an index where it's the prefix.

Reply:
> I only found queries for `device_code_signature` (primary key) and `request_id)` - do we need the challenge-id?

The challenge_id is not used as an index, the only reason it is there is to be uniform with the other `oauth2_token/code` tables. AFAICT it is not used there either, but thought maybe there is another reason for it.

> Also are we not missing a foreign key onto the oauth2 flow table (I'm not sure if we have that in the other tables)

Thought that there was some issue with adding that fk, but it seems to work now. Will try to add it.

> Since you're referencing `nid` as a foreign key, it will also need an index where it's the prefix.

Not sure why we need an index for that key. When will it ever be used? Do other tables have this index?

Reply:
> Thought that there was some issue with adding that fk, but it seems to work now. Will try to add it.

My bad, we need it for the foreign key.

> Not sure why we need an index for that key. When will it ever be used? Do other tables have this index?

It's needed because we have a foreign key. Any foreign key needs in (most?) databases also an index or otherwise CASCADE DELETE takes forever.

However, I think you can change

```
CREATE UNIQUE INDEX hydra_oauth2_device_auth_codes_user_code_signature_idx ON hydra_oauth2_device_auth_codes (user_code_signature, nid);
```

to


```
CREATE UNIQUE INDEX hydra_oauth2_device_auth_codes_user_code_signature_idx ON hydra_oauth2_device_auth_codes (nid, user_code_signature);
```

and then `nid` is covered and you don't need an additional index

Reply:
Understood, made `device_challenge` index unique and added a test to validate it (https://github.com/ory/hydra/pull/3912/commits/9e2e241ad8fd54912897f98bee5957330eeb9f5f). Also changed the user code index as you suggested to have `nid` first.